### ISSUE TITLE
Masker WAV-file after masking header chunk size does not match file size

### ISSUE THREAD

**Issue by wuddel** [Exmaralda-Org/exmaralda#280]
In a task of masking twice a WAV-file, i.e. after having once masked a fila and then using that masked file in a second run of masking, an error is thrown:
**File could not be masked:
Header chunk size (636641316) does not match file size (636641320)**
The Audio cannot be masked.
Having a look at the two files (original and masked) with sox, does not show any difference: 
![WAV-files_sox_comparison](https://user-images.githubusercontent.com/58472461/132486170-edd2ddf2-081d-4638-b54d-ef5073db0a48.png)
However, when looking with a RIFF viewer (cf. https://www.menasoft.com/blog/?p=34), the two WAV-files look rather different:
![WAV-files_riff_viewer_comparison](https://user-images.githubusercontent.com/58472461/132487154-5a5ad21a-4cb5-42d4-b0fe-465d6472a020.png)
Not only the ChunkSizes are different, but also the number of File-Infos (left panel), but this might not be relevant.

A workaround is currently to use an axternal program, e.g. VLC, to convert the masked audio to a new WAV-File with the same settings (sample rate, bit rate, etc.) of the original and load the result for the next round of masking into the Partitur Editor. The newly created file might be slightly different to the original, e.g. in file size, but that size is at least consistent with the ChunkSize in the WAV-file header.

**Comment by wuddel** [comment]
As the chunk size does not match the file size, while the samples before masking and after masking are identical, there must be some error in the generation of the chunk size in the WAV-file header.
The header is written in WavFile.java (cf. lines 118 - 124):
```
        // Set the main chunk size
        putLE(RIFF_CHUNK_ID, wavFile.buffer, 0, 4);
        putLE(mainChunkSize, wavFile.buffer, 4, 4);
        putLE(RIFF_TYPE_ID, wavFile.buffer, 8, 4);

        // Write out the header
        wavFile.oStream.write(wavFile.buffer, 0, 12);
```
Possibly, the error stems from the lines above (l. 103 - 107):
```
        long mainChunkSize = 4 + // Riff Type
                8 + // Format ID and size
                16 + // Format data
                8 + // Data ID and size
                dataChunkSize;
```
If we simply add 4 (for whatever reason), we would reach the same number as the file size.

**Comment by wuddel** [comment]
Recently, there was a Java Update. 
After checking on my Mac, where I haven't agreed on installing the Java Update, yet, I could mask an already masked WAV-File without the error message.

**Comment by berndmoos** [comment]
> Recently, there was a Java Update.
> After checking on my Mac, where I haven't agreed on installing the Java Update, yet, I could mask an already masked WAV-File without the error message.

All previews on Mac and Windows come with an integrated JVM, so they should not be affected by Java Updates. Are you using the "official" (not preview) version on the MAC?

**Comment by wuddel** [comment]
On Windows I'm using:

> Application version :1.7
> Build time :20210802.1543
> Operating system architecture : amd64
> OS version : 10.0
> JRE version : 14

On MAC I'm using:

> Application version :1.7
> Build time :20200302.0937
> Operating system : Mac OS X
> Operating system architecture : x86_64
> OS version : 10.15.7
> JRE version : 14

Oh, that's rather old!

I think that I always and only use the preview version. If I installed the official version, that would have been accidentally. I haven't found the terms "preview" or "official" in the log-data. Is there a way to check that?

**Comment by wuddel** [comment]
Comparing with hexdump the original WAV-file:
```
C:\Program Files\hexdump-2.0.2>hexdump -C N:\Workspace\FOLK\2021\FOLK_COAC_01\FOLK_COAC_01_A01.wav
000000  52 49 46 46 8c 64 f2 25 57 41 56 45 66 6d 74 20  RIFF.d.%WAVEfmt
000010  10 00 00 00 01 00 02 00 80 bb 00 00 00 ee 02 00  ................
000020  04 00 10 00 64 61 74 61 00 60 f2 25 00 00 00 00  ....data.`.%....
000030  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
000040  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
```

and the masked file:
```
C:\Program Files\hexdump-2.0.2>hexdump -C N:\Workspace\FOLK\2021\FOLK_COAC_01\FOLK_COAC_01_A01_mask1_Silence.WAV
000000  52 49 46 46 24 60 f2 25 57 41 56 45 66 6d 74 20  RIFF$`.%WAVEfmt
000010  10 00 00 00 01 00 02 00 80 bb 00 00 00 ee 02 00  ................
000020  04 00 10 00 64 61 74 61 00 60 f2 25 00 00 00 00  ....data.`.%....
000030  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
000040  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
```
the only difference at the header is between "RIFF" and "WAVE", i.e. the ChunkSize, according to:
![WAV-header](https://user-images.githubusercontent.com/58472461/133071451-cabd5645-73fd-4086-8c3e-23082140b2c9.png)

**Comment by wuddel** [comment]
The exception is thrown when a wav-File is opened: WavFile.openWavFile()
```
        // Check that the file size matches the number of bytes listed in header
        if (file.length() != chunkSize + 8) {
            throw new WavFileException("Header chunk size (" + chunkSize + ") does not match file size (" + file.length() + ")");
        }
```
however, the error message is slightly misleading, as it does not contain the "+ 8" from the if-statement. So: when the error message says that the chunk size indicated in the header (636641316), is 4 bytes smaller than the file size (636641320), the chunk size would need to be 4 bytes smaller than it currently is, for the if-statement to be wrong, i.e. matching sizes.

The following calculation, as it is done in WavFile.java seems to work:
```
dataChunkSize = wavFile.blockAlign * numFrames; // Sample 1 to Sample n
mainChunkSize = 
 4 + // Riff Type
 8 + // Format ID and size
16 + // Format data
 8 // Data ID and size
------------------------
36 // this is the chunk descriptor + fmt descriptor
+ 8 //beginning of data descriptor, i.e. "d a t a" + 4 bytes indicating the SubchunkSize
------------------------
44
+ dataChunkSize; // data starts from 44
```
In turn, the above reasoning could also mean that in our case, the file is 4 bytes too small, which would be - one sample?, i.e. one frame? Half a frame? This leads me to Masker.java, lines 50-55:
```
        // issue #269
        // Create a buffer of 1 frame
        lastBufferValues = new double[1 * numChannels];
        // Read frames into buffer
        wavFileIn.readFrames(lastBufferValues, 1);
```
As the numChannels is 2 (in a stereo signal), the "lastBufferValues" would be 2. When reading the next frame(s), the first one might be skipped...?
This explanation would be strange, however, as the output of sox says that the number of samples before and after masking are the same.

Since we haven't changed the code other than for issue #269, I guess that the problem stems from there.

**Comment by wuddel** [comment]
I think I found the problem:
`wavFileIn.readFrames(lastBufferValues, 1);`
With this line, we read the first frame, but we never write these samples into the wavFileOut.
I was not aware that the reading of samples continues throughout the process until the last sample. The decisive clue was that there is a method _getFramesRemaining()_ in _WavFile.java_, which shows this continuous frame-wise reading of samples.
As this line is already included in the constructor, the mismatch of the chunk and file-size in the output occurs independently from the method of masking (Silence, Brownian Noise, etc.).

What we could do for a solution, is either:

- close the _wavFileIn_ after reading the first frame and then reopen it for the remaining tasks, assuming that the reading of frames starts again from the very start

or

- after reading the first frame, writing it to the _wavFileOut_, hoping that the remaining tasks (masking, copying) are not affected.

Does that make sense?

**Comment by berndmoos** [comment]
Yes, that would also explain why the problem occurs with newer but not older (preview) versions -- the fix for #269 has only been there since June 2021.

The second solution sounds dodgy - we shouldn`t rely on hope. I forget what the line in the constructor is good for. It is probably true that closing and reopening the wav file will reset any pointers, so the first solution could work. Can you check that? Or -- depending on what the #269 lines are doing in the constructor, solve that differently?

**Comment by wuddel** [comment]
This is what I've just tested (solution 1):
```
        // issue #269
        // Create a buffer of 1 frame
        lastBufferValues = new double[1 * numChannels];
        // Read frames into buffer
        wavFileIn.readFrames(lastBufferValues, 1);
        // issue #280 fix of bugfix #269
        // as the first frame that was just read is never written, we need to 
        // re-load the wav file to reset all pointers
        wavFileIn.close(); // close
        wavFileIn = WavFile.openWavFile(fileIn); // re-open
```
I could mask an audio-file and mask it again without error message.

The following (solution 2) also worked without error message at two rounds of masking:
```
        // issue #269
        // Create a buffer of 1 frame
        lastBufferValues = new double[1 * numChannels];
        // Read frames into buffer
        int framesRead;
        framesRead = wavFileIn.readFrames(lastBufferValues, 1);
        // issue #280 second fix of bugfix #269
        wavFileOut.writeFrames(lastBufferValues, framesRead);
```

No other exception, e.g. as in #269, is thrown.

**Comment by berndmoos** [comment]
Should be fixed afaik
